File: m26-py/m26/elapsed_time.py
__author__ = 'cjoakim'

import math

from numbers import Number

from .constants import Constants


class ElapsedTime(object):

    def __init__(self, val):
        self.secs = 0
        self.hh = 0
        self.mm = 0
        self.ss = 0

        if not val:
            val = 0
        if isinstance(val, Number):
            self.initialize_from_number(val)
        elif isinstance(val, str):
            self.initialize_from_string(val)

    def initialize_from_number(self, val):
        sph = Constants.seconds_per_hour()
        self.secs = float(val)
        self.hh = math.floor(self.secs / sph)
        rem = self.secs - (self.hh * sph)
        self.mm = math.floor(rem / 60.0)
        self.ss = rem - (self.mm * 60.0)

    def initialize_from_string(self, val):
        stripped = str(val).strip()
        if len(stripped) > 0:
            tokens = stripped.split(':')
            if len(tokens) == 1:
                self.ss = self.to_float(tokens[0])
            elif len(tokens) == 2:
                self.mm = self.to_float(tokens[0])
                self.ss = self.to_float(tokens[1])
            elif len(tokens) == 3:
                self.hh = self.to_float(tokens[0])
                self.mm = self.to_float(tokens[1])
                self.ss = self.to_float(tokens[2])
            else:
                pass

        self.secs = (self.hh * 3600.0) + (self.mm * 60.0) + self.ss

    def to_float(self, s):
        try:
            return float(s)
        except ValueError:
            return float(0.0)

    def hours(self):
        return float(self.secs / Constants.seconds_per_hour())

    def as_hhmmss(self):
        hhs = self.zero_fill(self.hh)
        mms = self.zero_fill(self.mm)
        sss = self.zero_fill(self.ss)
        return "{0}:{1}:{2}".format(hhs, mms, sss)

    def zero_fill(self, n):
        if n < 10:
            return "0{0}".format(int(n))
        else:
            return "{0}".format(int(n))

    def __str__(self):
        template = "<ElapsedTime hh:{0} mm:{1} ss:{2} secs:{3}>"
        return template.format(self.hh, self.mm, self.ss, self.secs)

File: m26-js-ts/src/ElapsedTime.ts

import { Constants } from "./Constants";

export class ElapsedTime {
    
    arg  : string = null;
    secs : number = 0;
    hh   : number = 0;
    mm   : number = 0;
    ss   : number = 0;

    constructor(constructorArg: string = '0:00:00') {
        this.arg = constructorArg.trim();
        if (constructorArg.includes(':')) {
            // Initialize from a string
            try {
                let tokens = this.arg.split(':');
                if (tokens.length == 3) {
                    this.hh = parseInt(tokens[0], 10);
                    this.mm = parseInt(tokens[1], 10);
                    this.ss = parseInt(tokens[2], 10);
                }
                else if (tokens.length == 2) {
                    this.mm = parseInt(tokens[0], 10);
                    this.ss = parseInt(tokens[1], 10);
                }
                else if (tokens.length == 1) {
                    this.ss = parseInt(tokens[0], 10);
                }
                this.secs = (this.hh * 3600) + (this.mm * 60) + this.ss
            }
            catch (error) {
                console.log(error);
                this.resetValues();
            }
        }
        else {
            // Initialize from a number of seconds
            try {
                if (this.arg.length < 1) {
                    this.resetValues();
                }
                else {
                    this.secs = Number.parseInt(this.arg);
                    this.hh = Math.floor(this.secs / Constants.SECONDS_PER_HOUR);
                    let rem : number = this.secs - (this.hh * Constants.SECONDS_PER_HOUR)
                    this.mm = Math.floor(rem / 60.0);
                    this.ss = rem - (this.mm * 60.0);
                }
            }
            catch (error) {
                console.log(error);
                this.resetValues();
            }
        }
    }

    asHours() : number {
        return this.secs / Constants.SECONDS_PER_HOUR;
    }

    asHHMMSS() : string {
        if (this.secs < 1) {
            return '00:00:00';
        }
        let tokens = [];
        tokens.push(this.zeroPad(this.hh));
        tokens.push(this.zeroPad(this.mm));
        tokens.push(this.zeroPad(this.ss));
        return tokens.join(':');
    }

    zeroPad(n : number) : string {
        if (n < 10) {
            return '0' + n;
        }
        else {
            return '' + n;
        }
    }

    resetValues() : void {
        this.hh = 0;
        this.mm = 0;
        this.ss = 0;
        this.secs = 0;
    }
}
